client/2022-03-07/rfc-extensions-io.md

```diff
@@ -0,0 +1,51 @@
+# Extension System : IO RFC
+
+## Table of Contents
+1. [Introduction](#Introduction)
+2. [Filesystem Access](#Filesystem-Access)
+    - [API](#API)
+    - [Data Representation](#Data-Representation)
+3. [Networking](#Networking) 
+
+
+## Introduction
+
+Vignette maintains a comprehensive and robust API, inspired of VSCode. However, there are certain considerations to take into account. To design a successful extension system, we must take account of both security of the users and the flexibility we provide to the developers.
+
+## Filesystem Access
+
+Access to the file system is considered a dangerous permission to grant due to the dangers it present. As such the objective of the extension system is only to provide the following:
+
+* Do NOT grant host level filesystem access to the extension.
+* Present a limited API that covers common use-cases.
+* Do NOT allow the extension to manage OS-specific file-related attributes.
+
+### API
+
+Vignette's FS system will be a reduced version of the FS API of Node.js. A basic file operation may look something like this.
+
+```typescript
+import { vignette, fs } from 'vignette';
+
+export function activate() {
+    const fileBuf = fs.readFile('./data.txt');
+    if (typeof fileBuf === Blob)
+      vignette.notification.send('info', 'Successfully read data.txt');
+    else 
+      vignette.notification.send('error', `Expected Blob, got ${typeof fileBuf} instead.`)
+}
+```
+
+The filesystem API only allows basic file operations, `readFile`, `unlink`, and `writeFile` and only takes a `Buffer` or a `Blob` as a input.

```

Comment:
Upon checking, `Blob` is a part of the File API and is not part of the vanilla V8.

Reply:
Would you want to implement `Buffer` instead? We can represent this internally as a `IOStream`.